{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f4921055-bef4-4e82-ae9f-f167816e19c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import gc\n",
    "import time\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import xgboost as xgb\n",
    "import tables\n",
    "\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.linear_model import Ridge\n",
    "from sklearn.model_selection import train_test_split,cross_val_score\n",
    "from sklearn.metrics import make_scorer\n",
    "\n",
    "# Limit threads for numerical libraries to manage CPU usage\n",
    "os.environ[\"OPENBLAS_NUM_THREADS\"] = \"7\"\n",
    "os.environ[\"OMP_NUM_THREADS\"] = \"7\"\n",
    "\n",
    "base_dir = \"/home/skovtun/Python_projects/Kaggle/Single_cell/\"\n",
    "data_dir = os.path.join(base_dir, \"data\")\n",
    "random_state = 77\n",
    "\n",
    "os.chdir(data_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "43acfad6-1b11-446d-b556-88e746bd7818",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chrX',\n",
       "       'chr8', 'chr9', 'chr11', 'chr10', 'chr12', 'chr13', 'chr14',\n",
       "       'chr15', 'chr16', 'chr17', 'chr18', 'chr20', 'chr19', 'chrY',\n",
       "       'chr22', 'chr21', 'MT', 'KI270728.1', 'KI270727.1', 'KI270442.1',\n",
       "       'GL000225.1', 'GL000009.2', 'GL000194.1', 'GL000205.2',\n",
       "       'GL000195.1', 'KI270733.1', 'GL000219.1', 'GL000216.2',\n",
       "       'KI270744.1', 'KI270734.1', 'GL000213.1', 'GL000220.1',\n",
       "       'GL000218.1', 'KI270731.1', 'KI270750.1', 'KI270721.1',\n",
       "       'KI270726.1', 'KI270711.1', 'KI270713.1'], dtype=object)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Getting external file providing the mapping of human genes to their chromosome coordinates on \n",
    "#the GRCh38 genome to use for reducing number of columns for every target.\n",
    "# RAW LINE:\n",
    "# 1\thavana\tgene\t11869\t14409\t.\t+\t.\tgene_id \"ENSG00000223972\"; gene_version \"5\"; gene_name \"DDX11L1\"; gene_source \"havana\"; gene_biotype \"transcribed_unprocessed_pseudogene\";\n",
    "\n",
    "gtf_path = \"Homo_sapiens.GRCh38.98.gtf\"\n",
    "\n",
    "genes = []\n",
    "\n",
    "with open(gtf_path) as f:\n",
    "    for line in f:\n",
    "        #skipping comments\n",
    "        if line.startswith(\"#\"):\n",
    "            continue\n",
    "        #splitting the line\n",
    "        fields = line.strip().split(\"\\t\")\n",
    "        if fields[2] != \"gene\":\n",
    "            continue\n",
    "        \n",
    "        chrom = fields[0]\n",
    "        start = int(fields[3])\n",
    "        end = int(fields[4])\n",
    "        \n",
    "        attr = fields[8]\n",
    "        match_id = re.search(r'gene_id \"([^\"]+)\"',attr)\n",
    "        match_name = re.search(r'gene_name \"([^\"]+)\"',attr)\n",
    "        if match_id:\n",
    "            gene_id = match_id.group(1)\n",
    "        else:\n",
    "            continue\n",
    "        if match_name:\n",
    "            gene_name = match_name.group(1)\n",
    "        else:\n",
    "            gene_name = None\n",
    "        \n",
    "        genes.append([gene_id, gene_name, chrom, start, end])\n",
    "\n",
    "gene_df = pd.DataFrame(genes, columns=[\"gene_id\", \"gene_name\", \"chr\", \"start\", \"end\"])\n",
    "mapping = {str(i): f\"chr{i}\" for i in range(1,23)}\n",
    "mapping['X'] = 'chrX'\n",
    "mapping['Y'] = 'chrY'\n",
    "gene_df['chr'] = gene_df['chr'].map(mapping).fillna(gene_df['chr'])\n",
    "gene_df['chr'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3bcbb224-4875-400d-becf-0da414d0c8d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(60623, 5)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "3695775a-c526-40d2-aa09-1b2396508965",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(23404, 4)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Getting a list of genes and all coordinates, parsing ATAC Peaks\n",
    "path = \"train_multi_inputs.h5\"\n",
    "cols = pd.read_hdf(path, key=\"train_multi_inputs\", start=0, stop=1).columns\n",
    "path = \"train_multi_targets.h5\"\n",
    "cols_t = pd.read_hdf(path, key = 'train_multi_targets', start = 0, stop=1).columns\n",
    "\n",
    "#From the column names of the multi creating a dataframe with location name, start and end.\n",
    "arr = pd.Series(cols.values)\n",
    "r_p = r'([^:]+):([\\d]+)-([\\d]+)'\n",
    "chr_ranges = arr.str.extract(r_p)\n",
    "chr_ranges.columns = ['chr','start','end']\n",
    "chr_ranges['start'] = chr_ranges['start'].astype(int)\n",
    "chr_ranges['end'] = chr_ranges['end'].astype(int)\n",
    "\n",
    "#Reducing gene_df by choosing onnly chr present in chr_ranges and only gene_id's present as targets.\n",
    "multi_chr = list(chr_ranges['chr'].unique())\n",
    "gene_df_multi = gene_df[gene_df['chr'].isin(multi_chr)]\n",
    "gene_df_multi = gene_df_multi.set_index('gene_id')\n",
    "missing = cols_t.difference(gene_df_multi.index)\n",
    "gene_df_multi = gene_df_multi.loc[gene_df_multi.index.intersection(cols_t)]\n",
    "gene_df_multi.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "08f0065f-e88b-4b24-bc4f-2bdaebd109be",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculating maximum amount of features for every target.\n",
    "window = 200000\n",
    "\n",
    "genes = gene_df_multi.loc[gene_df_multi.index.intersection(cols_t)].copy()\n",
    "genes[\"left\"]  = genes[\"start\"] - window\n",
    "genes[\"right\"] = genes[\"end\"]   + window\n",
    "genes[\"gene_id\"] = genes.index  # preserve gene_id as a column\n",
    "\n",
    "chr_ranges = chr_ranges.rename(columns={\n",
    "    \"start\": \"start_peak\",\n",
    "    \"end\":   \"end_peak\"\n",
    "})\n",
    "\n",
    "target_cols = pd.DataFrame()\n",
    "results = []\n",
    "gene_id = []\n",
    "start_peak = []\n",
    "end_peak = []\n",
    "\n",
    "for chr_name, chr_peaks in chr_ranges.groupby(\"chr\"):\n",
    "    sub_genes = genes[genes[\"chr\"] == chr_name]\n",
    "    if sub_genes.empty or chr_peaks.empty:\n",
    "        continue\n",
    "\n",
    "    merged = sub_genes[['gene_id', 'chr', 'left', 'right']].merge(chr_peaks[['chr', 'start_peak', 'end_peak']], on=\"chr\")\n",
    "\n",
    "    mask = (\n",
    "        (merged[\"start_peak\"] >= merged[\"left\"]) &\n",
    "        (merged[\"end_peak\"]   <= merged[\"right\"])\n",
    "    )\n",
    "    m = merged[mask]\n",
    "    results.append(m)\n",
    "        \n",
    "target_cols = pd.concat(results)\n",
    "target_cols['col'] = target_cols['chr'].astype(str)+\":\"+target_cols['start_peak'].astype(str)+\"-\"+target_cols['end_peak'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f13de7ed-3f49-41ec-8048-8d1227286e92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1376144, 7)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_cols.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1aab5075-9d5d-49bc-8d38-f1466cbc64cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "target_cols.to_parquet(\"target_cols.parquet\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66903b7e-19ec-4d98-ad55-97e42c6d3859",
   "metadata": {},
   "source": [
    "For every target we have from 1 to 234 features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "c49f8a87-4bf5-429e-b670-4393f8eeb063",
   "metadata": {},
   "outputs": [],
   "source": [
    "with tables.open_file(\"train_multi_inputs.h5\", \"r\") as f:\n",
    "    values = f.get_node(\"/train_multi_inputs/axis1\")[:]\n",
    "valid_ids = np.char.decode(values, encoding = 'utf-8')\n",
    "\n",
    "metadata_old = pd.read_csv('metadata.csv', index_col = 'cell_id')\n",
    "fix = pd.read_csv('metadata_cite_day_2_donor_27678.csv', index_col = 'cell_id')\n",
    "metadata = pd.concat([metadata_old, fix], axis = 0)\n",
    "del metadata_old, fix\n",
    "multi_metadata  = metadata.loc[train_id,['day','donor']]\n",
    "del metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "2d7f0b4d-83c4-4b92-baa9-7e1d43d1b4de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- Data Splitting and Feature Separation ---\n",
    "# Random Train-Test Split (80/20)\n",
    "random_state=77\n",
    "rng = np.random.default_rng(random_state)\n",
    "shuffled_idx = rng.permutation(valid_ids)\n",
    "\n",
    "split_point = int(len(shuffled_idx) * 0.8)\n",
    "train_idx = shuffled_idx[:split_point]\n",
    "test_idx = shuffled_idx[split_point:]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f061fce8-4788-446c-873f-4cfbe4df9697",
   "metadata": {},
   "source": [
    "Writing a pipeline that will be computed for each of 23300 targets\n",
    "First step: getting the data.\n",
    "Genes and peaks correspondence are in target_cols\n",
    "For now working on how to do everything for just one target.\n",
    "First decision: reading all targets at once. array Y, with columns stored separately as items_Y.\n",
    "Now getting peak values in batches: for 256 genes at once."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "96fcda33-1cf8-454c-a1db-94ce1b7ab649",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chr                             chr19\n",
       "gene_id               ENSG00000104848\n",
       "start_peak                   49084891\n",
       "end_peak                     49085777\n",
       "col_t         chr19:49084891-49085777\n",
       "col           chr19:49084891-49085777\n",
       "Name: 8867055, dtype: object"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_cols.iloc[728357]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a1f077d2-ebf9-49a8-b5e8-fa69f297ed11",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_id = 'ENSG00000104848'\n",
    "list_of_peak_strings = list(target_cols[target_cols['gene_id'] == gene_id]['col'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "45bee74c-593c-4f0a-9298-42f3e9db0090",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.30817723274230957\n"
     ]
    }
   ],
   "source": [
    "import tables\n",
    "import numpy as np\n",
    "import time\n",
    "start = time.time()\n",
    "with tables.open_file(\"train_multi_targets.h5\") as f:\n",
    "    items_Y = f.root.train_multi_targets.block0_items[:].astype(str)  # gene_ids\n",
    "    arr_Y   = f.root.train_multi_targets.block0_values                 # (n_cells, n_genes)\n",
    "    Y = np.array(arr_Y, dtype=np.float32)                              # full target in RAM\n",
    "print((time.time() - start)/60)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d55afc92-7a0f-4411-bfed-2ea96b79f7ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ENSG00000121410', 'ENSG00000268895', 'ENSG00000175899',\n",
       "       'ENSG00000245105', 'ENSG00000166535'], dtype='<U15')"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "items_Y[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a4601b2a-678f-4306-aa25-a69460cbc923",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.       , 0.       , 0.       , 0.       , 0.       ],\n",
       "       [0.       , 0.       , 0.       , 0.       , 0.       ],\n",
       "       [0.       , 0.       , 0.       , 0.       , 0.       ],\n",
       "       [0.       , 4.5079365, 0.       , 0.       , 0.       ],\n",
       "       [0.       , 0.       , 0.       , 0.       , 0.       ]],\n",
       "      dtype=float32)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Y[:5,:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a8773856-ad35-463a-8058-bdd4f98bf319",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[39079, 50, 51, 32831, 64]\n"
     ]
    }
   ],
   "source": [
    "#Getting a list of genes and all coordinates\n",
    "# path = \"train_multi_inputs.h5\"\n",
    "# cols = pd.read_hdf(path, key=\"train_multi_inputs\", start=0, stop=1).columns\n",
    "# path = \"train_multi_targets.h5\"\n",
    "# cols_t = pd.read_hdf(path, key = 'train_multi_targets', start = 0, stop=1).columns\n",
    "\n",
    "peak_to_idx = {name: i for i, name in enumerate(cols)}\n",
    "gene_to_idx = {name: i for i, name in enumerate(items_Y)}\n",
    "gene_list = items_Y[:256]\n",
    "list_of_peak_strings = list(target_cols[target_cols['gene_id'].isin(gene_list)]['col'])\n",
    "selected_peaks_id = list(set([peak_to_idx[name] for name in list_of_peak_strings]))\n",
    "print(selected_peaks_id[:5])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0c9ab634-96f4-403b-b7bd-095cf3a668cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.2264537493387857\n"
     ]
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "col_idx =  selected_peaks_id    # your indices\n",
    "batch = 2000\n",
    "\n",
    "with tables.open_file(\"train_multi_inputs.h5\", \"r\") as f:\n",
    "    values = f.get_node(\"/train_multi_inputs/block0_values\")\n",
    "    items  = f.get_node(\"/train_multi_inputs/block0_items\")[:]  # fully read\n",
    "\n",
    "    n_rows = values.shape[0]\n",
    "    n_cols = len(col_idx)\n",
    "\n",
    "    arr = np.empty((n_rows, n_cols), dtype=values.dtype)\n",
    "\n",
    "    for start in range(0, n_rows, batch):\n",
    "        end = min(start + batch, n_rows)\n",
    "        block = values[start:end, :]        # chunk-aligned read\n",
    "        arr[start:end, :] = block[:, col_idx]\n",
    "\n",
    "# now the file is closed; items is safe\n",
    "df = pd.DataFrame(arr, columns=items[col_idx])\n",
    "df.columns = df.columns.str.decode(\"utf-8\")\n",
    "print((time.time() - start_time)/60)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "566753c9-6ad7-4c75-bab1-ad357b2e3344",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr13:95491683-95492594</th>\n",
       "      <th>KI270721.1:2091-2978</th>\n",
       "      <th>KI270721.1:8381-9254</th>\n",
       "      <th>chr12:942761-943471</th>\n",
       "      <th>chr10:100001240-100002159</th>\n",
       "      <th>chr10:100006137-100006993</th>\n",
       "      <th>chr10:100009448-100010341</th>\n",
       "      <th>chr10:100013840-100014756</th>\n",
       "      <th>chr10:100043298-100044049</th>\n",
       "      <th>chr1:33111718-33112623</th>\n",
       "      <th>...</th>\n",
       "      <th>chr7:20906021-20906882</th>\n",
       "      <th>chr7:20910868-20911751</th>\n",
       "      <th>chr7:20913372-20914288</th>\n",
       "      <th>chr13:95463279-95464129</th>\n",
       "      <th>chr7:20926434-20927110</th>\n",
       "      <th>chr7:20927658-20928561</th>\n",
       "      <th>chr13:95474906-95475770</th>\n",
       "      <th>chr13:95476873-95477605</th>\n",
       "      <th>chr13:95478266-95479176</th>\n",
       "      <th>chr13:95479928-95480825</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.213161</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.118048</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.544025</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.103673</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 11506 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr13:95491683-95492594  KI270721.1:2091-2978  KI270721.1:8381-9254  \\\n",
       "0                      0.0                   0.0                   0.0   \n",
       "1                      0.0                   0.0                   0.0   \n",
       "2                      0.0                   0.0                   0.0   \n",
       "3                      0.0                   0.0                   0.0   \n",
       "4                      0.0                   0.0                   0.0   \n",
       "\n",
       "   chr12:942761-943471  chr10:100001240-100002159  chr10:100006137-100006993  \\\n",
       "0                  0.0                        0.0                        0.0   \n",
       "1                  0.0                        0.0                        0.0   \n",
       "2                  0.0                        0.0                        0.0   \n",
       "3                  0.0                        0.0                        0.0   \n",
       "4                  0.0                        0.0                        0.0   \n",
       "\n",
       "   chr10:100009448-100010341  chr10:100013840-100014756  \\\n",
       "0                   0.000000                        0.0   \n",
       "1                   0.000000                        0.0   \n",
       "2                   0.000000                        0.0   \n",
       "3                   1.544025                        0.0   \n",
       "4                   0.000000                        0.0   \n",
       "\n",
       "   chr10:100043298-100044049  chr1:33111718-33112623  ...  \\\n",
       "0                        0.0                0.000000  ...   \n",
       "1                        0.0                0.000000  ...   \n",
       "2                        0.0                0.000000  ...   \n",
       "3                        0.0                4.103673  ...   \n",
       "4                        0.0                0.000000  ...   \n",
       "\n",
       "   chr7:20906021-20906882  chr7:20910868-20911751  chr7:20913372-20914288  \\\n",
       "0                     0.0                     0.0                     0.0   \n",
       "1                     0.0                     0.0                     0.0   \n",
       "2                     0.0                     0.0                     0.0   \n",
       "3                     0.0                     0.0                     0.0   \n",
       "4                     0.0                     0.0                     0.0   \n",
       "\n",
       "   chr13:95463279-95464129  chr7:20926434-20927110  chr7:20927658-20928561  \\\n",
       "0                 5.213161                     0.0                     0.0   \n",
       "1                 0.000000                     0.0                     0.0   \n",
       "2                 0.000000                     0.0                     0.0   \n",
       "3                 0.000000                     0.0                     0.0   \n",
       "4                 0.000000                     0.0                     0.0   \n",
       "\n",
       "   chr13:95474906-95475770  chr13:95476873-95477605  chr13:95478266-95479176  \\\n",
       "0                      0.0                 0.000000                      0.0   \n",
       "1                      0.0                 0.000000                      0.0   \n",
       "2                      0.0                 5.118048                      0.0   \n",
       "3                      0.0                 0.000000                      0.0   \n",
       "4                      0.0                 0.000000                      0.0   \n",
       "\n",
       "   chr13:95479928-95480825  \n",
       "0                      0.0  \n",
       "1                      0.0  \n",
       "2                      0.0  \n",
       "3                      0.0  \n",
       "4                      0.0  \n",
       "\n",
       "[5 rows x 11506 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "905dd452-103b-4378-b9d0-2267978de323",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.linear_model import Ridge\n",
    "import pandas as pd\n",
    "\n",
    "corr_max = []\n",
    "y_train_var = []\n",
    "y_test_var = []\n",
    "corr_c = []\n",
    "\n",
    "for i in range(256):\n",
    "    gene_id = items_Y[i]\n",
    "\n",
    "    # pick peaks for this gene\n",
    "    peaks_str = list(target_cols.loc[target_cols['gene_id'] == gene_id, 'col'])\n",
    "\n",
    "    # skip if no peaks\n",
    "    if len(peaks_str) == 0:\n",
    "        corr_max.append(np.nan)\n",
    "        y_train_var.append(np.nan)\n",
    "        y_test_var.append(np.nan)\n",
    "        corr_c.append(np.nan)\n",
    "        continue\n",
    "\n",
    "    df_gene = df[peaks_str]\n",
    "\n",
    "    # get expression for this gene\n",
    "    y_gene = Y[:, i]\n",
    "\n",
    "    # 1) max absolute peak–gene correlation\n",
    "    y_series = pd.Series(y_gene, index=df_gene.index)\n",
    "    corrs = df_gene.corrwith(y_series)\n",
    "    corr_max.append(corrs.abs().max())\n",
    "\n",
    "    # 2) train/test split\n",
    "    X_train, X_test, y_train, y_test = train_test_split(\n",
    "        df_gene, y_gene, train_size=0.8, random_state=77\n",
    "    )\n",
    "\n",
    "    # 3) scale\n",
    "    scaler = StandardScaler(with_mean=False)\n",
    "    X_train_scaled = scaler.fit_transform(X_train)\n",
    "    X_test_scaled = scaler.transform(X_test)\n",
    "\n",
    "    # store variances\n",
    "    y_train_var.append(np.var(y_train))\n",
    "    y_test_var.append(np.var(y_test))\n",
    "\n",
    "    # 4) ridge + test correlation\n",
    "    model = Ridge(alpha=1.0, fit_intercept=True)\n",
    "    model.fit(X_train_scaled, y_train)\n",
    "    y_pred = model.predict(X_test_scaled)\n",
    "    corr_c.append(np.corrcoef(y_test, y_pred)[0, 1])\n",
    "summary = pd.DataFrame({\n",
    "    \"gene_id\": items_Y[:256],\n",
    "    \"max_peak_corr\": corr_max,\n",
    "    \"y_train_var\": y_train_var,\n",
    "    \"y_test_var\": y_test_var,\n",
    "    \"ridge_test_r\": corr_c,\n",
    "})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "22289f73-4c52-4795-9cbe-35359fa2ee80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import gc\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2c4403c9-99dd-4d54-aec9-eaed836bcc66",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             gene_id  max_peak_corr  y_train_var  y_test_var  ridge_test_r\n",
      "13   ENSG00000115977       0.071275     6.815775    6.795776      0.139281\n",
      "34   ENSG00000107331       0.056327     4.622707    4.605511      0.114302\n",
      "48   ENSG00000115657       0.076905     1.428170    1.467709      0.148674\n",
      "56   ENSG00000108846       0.084350     0.644879    0.687800      0.137206\n",
      "57   ENSG00000125257       0.112532     8.042895    8.034110      0.283323\n",
      "87   ENSG00000136379       0.061913     2.587219    2.633019      0.102538\n",
      "102  ENSG00000099204       0.073839     6.154472    6.195627      0.136348\n",
      "105  ENSG00000175164       0.050723     5.164624    5.191978      0.102156\n",
      "113  ENSG00000166016       0.056345     2.810173    2.714567      0.138496\n",
      "224  ENSG00000225792       0.065270     2.034776    1.949889      0.115125\n",
      "240  ENSG00000265206       0.070217     6.259258    6.303818      0.129724\n"
     ]
    }
   ],
   "source": [
    "good = summary[\n",
    "    (summary[\"max_peak_corr\"].abs() > 0.05) &\n",
    "    (summary[\"ridge_test_r\"] > 0.1) &\n",
    "    (summary[\"y_train_var\"] > 0.5)\n",
    "]\n",
    "print(good)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "8c20057c-448b-4e5a-9a41-3fa7962538e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENSG00000115977', 'ENSG00000107331', 'ENSG00000115657', 'ENSG00000108846', 'ENSG00000125257', 'ENSG00000136379', 'ENSG00000099204', 'ENSG00000175164', 'ENSG00000166016', 'ENSG00000225792', 'ENSG00000265206']\n"
     ]
    }
   ],
   "source": [
    "print(list(good['gene_id']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "a0a6280a-ea0d-4393-87c6-bb47c6e4ca9f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best alpha: 10000.0\n",
      "Best alpha: 10000.0\n",
      "Best alpha: 10000.0\n",
      "Best alpha: 10000.0\n",
      "Best alpha: 10000.0\n",
      "Best alpha: 10000.0\n",
      "Best alpha: 10000.0\n",
      "Best alpha: 10000.0\n",
      "Best alpha: 10000.0\n",
      "Best alpha: 10000.0\n",
      "Best alpha: 10000.0\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.linear_model import Ridge\n",
    "import pandas as pd\n",
    "from sklearn.linear_model import RidgeCV\n",
    "from sklearn.metrics import make_scorer\n",
    "\n",
    "corr_max = []\n",
    "y_train_var = []\n",
    "y_test_var = []\n",
    "corr_c = []\n",
    "corr_c_t = []\n",
    "\n",
    "def pearsonr_score(y_true, y_pred):\n",
    "    return np.corrcoef(y_true, y_pred)[0,1]\n",
    "\n",
    "pearson_scorer = make_scorer(pearsonr_score, greater_is_better=True)\n",
    "\n",
    "for gene_id in good_gene_id:\n",
    "\n",
    "    # pick peaks for this gene\n",
    "    peaks_str = list(target_cols.loc[target_cols['gene_id'] == gene_id, 'col'])\n",
    "\n",
    "    # skip if no peaks\n",
    "    if len(peaks_str) == 0:\n",
    "        corr_max.append(np.nan)\n",
    "        y_train_var.append(np.nan)\n",
    "        y_test_var.append(np.nan)\n",
    "        corr_c.append(np.nan)\n",
    "        corr_c_t.append(np.nan)\n",
    "        continue\n",
    "\n",
    "    df_gene = df[peaks_str]\n",
    "    \n",
    "    # get expression for this gene\n",
    "    y_gene = Y[:,good_gene_id[good_gene_id == gene_id].index[0]]\n",
    "\n",
    "    # 1) max absolute peak–gene correlation\n",
    "    y_series = pd.Series(y_gene, index=df_gene.index)\n",
    "    corrs = df_gene.corrwith(y_series)\n",
    "    corr_max.append(corrs.abs().max())\n",
    "\n",
    "    # 2) train/test split\n",
    "    X_train, X_test, y_train, y_test = train_test_split(\n",
    "        df_gene, y_gene, train_size=0.8, random_state=77\n",
    "    )\n",
    "\n",
    "    # 3) scale\n",
    "    scaler = StandardScaler(with_mean=False)\n",
    "    X_train_scaled = scaler.fit_transform(X_train)\n",
    "    X_test_scaled = scaler.transform(X_test)\n",
    "\n",
    "    # store variances\n",
    "    y_train_var.append(np.var(y_train))\n",
    "    y_test_var.append(np.var(y_test))\n",
    "\n",
    "    # 4) ridge + test correlation\n",
    "    model = Ridge(alpha=1.0, fit_intercept=True)\n",
    "    model.fit(X_train_scaled, y_train)\n",
    "    y_pred = model.predict(X_test_scaled)\n",
    "    corr_c.append(np.corrcoef(y_test, y_pred)[0, 1])\n",
    "\n",
    "     # 5) ridge + tuning alpha\n",
    "    alphas = np.logspace(-4, 4, 100)\n",
    "\n",
    "    model = RidgeCV(alphas=alphas, scoring=pearson_scorer)\n",
    "    model.fit(X_train_scaled, y_train)\n",
    "    \n",
    "    best_alpha = model.alpha_\n",
    "    print(\"Best alpha:\", best_alpha)\n",
    "        \n",
    "    model = Ridge(alpha=best_alpha, fit_intercept=True)\n",
    "    model.fit(X_train_scaled, y_train)\n",
    "    y_pred = model.predict(X_test_scaled)\n",
    "    corr_c_t.append(np.corrcoef(y_test, y_pred)[0, 1])\n",
    "\n",
    "summary = pd.DataFrame({\n",
    "    \"gene_id\": good_gene_id,\n",
    "    \"max_peak_corr\": corr_max,\n",
    "    \"y_train_var\": y_train_var,\n",
    "    \"y_test_var\": y_test_var,\n",
    "    \"ridge_test_r\": corr_c,\n",
    "    \"ridge_test_rt\": corr_c_t,\n",
    "})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "4e104e0f-1505-4cf4-91f9-be413db3c6da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             gene_id  max_peak_corr  y_train_var  y_test_var  ridge_test_r  \\\n",
      "13   ENSG00000115977       0.071275     6.815775    6.795776      0.139281   \n",
      "34   ENSG00000107331       0.056327     4.622707    4.605511      0.114302   \n",
      "48   ENSG00000115657       0.076905     1.428170    1.467709      0.148674   \n",
      "56   ENSG00000108846       0.084350     0.644879    0.687800      0.137206   \n",
      "57   ENSG00000125257       0.112532     8.042895    8.034110      0.283323   \n",
      "87   ENSG00000136379       0.061913     2.587219    2.633019      0.102538   \n",
      "102  ENSG00000099204       0.073839     6.154472    6.195627      0.136348   \n",
      "105  ENSG00000175164       0.050723     5.164624    5.191978      0.102156   \n",
      "113  ENSG00000166016       0.056345     2.810173    2.714567      0.138496   \n",
      "224  ENSG00000225792       0.065270     2.034776    1.949889      0.115125   \n",
      "240  ENSG00000265206       0.070217     6.259258    6.303818      0.129724   \n",
      "\n",
      "     ridge_test_rt  \n",
      "13        0.139407  \n",
      "34        0.114481  \n",
      "48        0.148822  \n",
      "56        0.137329  \n",
      "57        0.283405  \n",
      "87        0.102589  \n",
      "102       0.136405  \n",
      "105       0.102365  \n",
      "113       0.138575  \n",
      "224       0.115194  \n",
      "240       0.129955  \n"
     ]
    }
   ],
   "source": [
    "print(summary)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcd57703-634c-4221-8614-6055e18319c4",
   "metadata": {},
   "source": [
    "Tuning Ridge didn't help, let's try other models that could help with capturing non linearity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0bd7a119-80ff-4edb-9352-052b246154b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index([13, 34, 48, 56, 57, 87, 102, 105, 113, 224, 240], dtype='int64')"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "good['gene_id'].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "995e41d9-05db-40ae-a98d-e4709dc58a32",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best params: {'alpha': 1, 'gamma': 0.001, 'kernel': 'rbf'}\n",
      "Best params: {'alpha': 0.1, 'gamma': 0.001, 'kernel': 'rbf'}\n",
      "Best params: {'alpha': 1, 'gamma': 0.001, 'kernel': 'rbf'}\n",
      "Best params: {'alpha': 10, 'gamma': 0.001, 'kernel': 'rbf'}\n",
      "Best params: {'alpha': 1, 'gamma': 0.001, 'kernel': 'rbf'}\n",
      "Best params: {'alpha': 10, 'gamma': 0.001, 'kernel': 'rbf'}\n",
      "Best params: {'alpha': 1, 'gamma': 0.001, 'kernel': 'rbf'}\n",
      "Best params: {'alpha': 1, 'gamma': 0.001, 'kernel': 'rbf'}\n",
      "Best params: {'alpha': 10, 'gamma': 0.001, 'kernel': 'rbf'}\n",
      "Best params: {'alpha': 1, 'gamma': 0.001, 'kernel': 'rbf'}\n",
      "Best params: {'alpha': 1, 'gamma': 0.001, 'kernel': 'rbf'}\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.linear_model import Ridge\n",
    "import pandas as pd\n",
    "from sklearn.metrics import make_scorer\n",
    "from sklearn.kernel_ridge import KernelRidge\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "\n",
    "corr_max = []\n",
    "y_train_var = []\n",
    "y_test_var = []\n",
    "corr_c = []\n",
    "corr_c_t = []\n",
    "\n",
    "def pearsonr_score(y_true, y_pred):\n",
    "    return np.corrcoef(y_true, y_pred)[0,1]\n",
    "\n",
    "pearson_scorer = make_scorer(pearsonr_score, greater_is_better=True)\n",
    "good_gene_id = good['gene_id']\n",
    "for gene_id in good_gene_id:\n",
    "\n",
    "    # pick peaks for this gene\n",
    "    peaks_str = list(target_cols.loc[target_cols['gene_id'] == gene_id, 'col'])\n",
    "\n",
    "    # skip if no peaks\n",
    "    if len(peaks_str) == 0:\n",
    "        corr_max.append(np.nan)\n",
    "        y_train_var.append(np.nan)\n",
    "        y_test_var.append(np.nan)\n",
    "        corr_c.append(np.nan)\n",
    "        corr_c_t.append(np.nan)\n",
    "        continue\n",
    "\n",
    "    df_gene = df[peaks_str]\n",
    "    \n",
    "    # get expression for this gene\n",
    "    y_gene = Y[:,good_gene_id[good_gene_id == gene_id].index[0]]\n",
    "\n",
    "    # 1) max absolute peak–gene correlation\n",
    "    y_series = pd.Series(y_gene, index=df_gene.index)\n",
    "    corrs = df_gene.corrwith(y_series)\n",
    "    corr_max.append(corrs.abs().max())\n",
    "\n",
    "    # 2) train/test split, putting limitation on train size because of memory limitation\n",
    "    X_train, X_test, y_train, y_test = train_test_split(\n",
    "        df_gene, y_gene, train_size=10000, random_state=77\n",
    "    )\n",
    "\n",
    "    # 3) scale\n",
    "    scaler = StandardScaler(with_mean=False)\n",
    "    X_train_scaled = scaler.fit_transform(X_train)\n",
    "    X_test_scaled = scaler.transform(X_test)\n",
    "\n",
    "    # store variances\n",
    "    y_train_var.append(np.var(y_train))\n",
    "    y_test_var.append(np.var(y_test))\n",
    "\n",
    "    # 4) ridge + test correlation\n",
    "    model = Ridge(alpha=1.0, fit_intercept=True)\n",
    "    model.fit(X_train_scaled, y_train)\n",
    "    y_pred = model.predict(X_test_scaled)\n",
    "    corr_c.append(np.corrcoef(y_test, y_pred)[0, 1])\n",
    "\n",
    "    # 5) Tuning KernelRidgee + test correlation\n",
    "    param_grid = {\n",
    "        \"alpha\": [0.1, 1, 10],\n",
    "        \"gamma\": [1e-3, 1e-2],\n",
    "        \"kernel\": [\"rbf\"],\n",
    "    }\n",
    "    \n",
    "    kr = KernelRidge()\n",
    "    \n",
    "    grid = GridSearchCV(\n",
    "        kr,\n",
    "        param_grid,\n",
    "        scoring=pearson_scorer,\n",
    "        cv=3,\n",
    "        n_jobs=-1\n",
    "    )\n",
    "\n",
    "    grid.fit(X_train_scaled, y_train)\n",
    "    \n",
    "    print(\"Best params:\", grid.best_params_)\n",
    "    best_model = grid.best_estimator_\n",
    "\n",
    "    y_pred = best_model.predict(X_test_scaled)\n",
    "    corr_c_t.append(np.corrcoef(y_test, y_pred)[0, 1])\n",
    "\n",
    "summary = pd.DataFrame({\n",
    "    \"gene_id\": good_gene_id,\n",
    "    \"max_peak_corr\": corr_max,\n",
    "    \"y_train_var\": y_train_var,\n",
    "    \"y_test_var\": y_test_var,\n",
    "    \"ridge_test_r\": corr_c,\n",
    "    \"kenel_ridge_test_rt\": corr_c_t,\n",
    "})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3d39e160-56d8-4e8a-abca-848224cb9bc6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             gene_id  max_peak_corr  y_train_var  y_test_var  ridge_test_r  \\\n",
      "13   ENSG00000115977       0.071275     6.765996    6.816464      0.122643   \n",
      "34   ENSG00000107331       0.056327     4.655192    4.615505      0.085861   \n",
      "48   ENSG00000115657       0.076905     1.487994    1.430646      0.111842   \n",
      "56   ENSG00000108846       0.084350     0.648581    0.653983      0.106797   \n",
      "57   ENSG00000125257       0.112532     7.976235    8.047922      0.277203   \n",
      "87   ENSG00000136379       0.061913     2.511168    2.605215      0.079718   \n",
      "102  ENSG00000099204       0.073839     6.094682    6.169738      0.115238   \n",
      "105  ENSG00000175164       0.050723     5.195562    5.167503      0.084072   \n",
      "113  ENSG00000166016       0.056345     2.841272    2.785895      0.110262   \n",
      "224  ENSG00000225792       0.065270     2.106782    2.008544      0.097647   \n",
      "240  ENSG00000265206       0.070217     6.246537    6.270519      0.114096   \n",
      "\n",
      "     kenel_ridge_test_rt  \n",
      "13              0.129831  \n",
      "34              0.084540  \n",
      "48              0.113091  \n",
      "56              0.104395  \n",
      "57              0.272211  \n",
      "87              0.082594  \n",
      "102             0.115726  \n",
      "105             0.086151  \n",
      "113             0.123956  \n",
      "224             0.098624  \n",
      "240             0.112537  \n"
     ]
    }
   ],
   "source": [
    "print(summary)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c98f4c1d-784c-4d6a-868a-caa9e39038e3",
   "metadata": {},
   "source": [
    "Minimal gains, meaning most of the signal was captured by Ridge (linear)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "672c9eb5-57aa-42fe-9e01-351115c12356",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.linear_model import Ridge\n",
    "import pandas as pd\n",
    "from sklearn.metrics import make_scorer\n",
    "import lightgbm as lgb\n",
    "\n",
    "\n",
    "\n",
    "corr_max = []\n",
    "y_train_var = []\n",
    "y_test_var = []\n",
    "corr_c = []\n",
    "corr_c_t = []\n",
    "\n",
    "def pearsonr_score(y_true, y_pred):\n",
    "    return np.corrcoef(y_true, y_pred)[0,1]\n",
    "\n",
    "pearson_scorer = make_scorer(pearsonr_score, greater_is_better=True)\n",
    "good_gene_id = good['gene_id']\n",
    "for gene_id in good_gene_id:\n",
    "\n",
    "    # pick peaks for this gene\n",
    "    peaks_str = list(target_cols.loc[target_cols['gene_id'] == gene_id, 'col'])\n",
    "\n",
    "    # skip if no peaks\n",
    "    if len(peaks_str) == 0:\n",
    "        corr_max.append(np.nan)\n",
    "        y_train_var.append(np.nan)\n",
    "        y_test_var.append(np.nan)\n",
    "        corr_c.append(np.nan)\n",
    "        corr_c_t.append(np.nan)\n",
    "        continue\n",
    "\n",
    "    df_gene = df[peaks_str]\n",
    "    \n",
    "    # get expression for this gene\n",
    "    y_gene = Y[:,good_gene_id[good_gene_id == gene_id].index[0]]\n",
    "\n",
    "    # 1) max absolute peak–gene correlation\n",
    "    y_series = pd.Series(y_gene, index=df_gene.index)\n",
    "    corrs = df_gene.corrwith(y_series)\n",
    "    corr_max.append(corrs.abs().max())\n",
    "\n",
    "    # 2) train/test split, putting limitation on train size because of memory limitation\n",
    "    X_train, X_test, y_train, y_test = train_test_split(\n",
    "        df_gene, y_gene, train_size=0.8, random_state=77\n",
    "    )\n",
    "\n",
    "    # 3) scale\n",
    "    scaler = StandardScaler(with_mean=False)\n",
    "    X_train_scaled = scaler.fit_transform(X_train)\n",
    "    X_test_scaled = scaler.transform(X_test)\n",
    "\n",
    "    # store variances\n",
    "    y_train_var.append(np.var(y_train))\n",
    "    y_test_var.append(np.var(y_test))\n",
    "\n",
    "    # 4) ridge + test correlation\n",
    "    model = Ridge(alpha=1.0, fit_intercept=True)\n",
    "    model.fit(X_train_scaled, y_train)\n",
    "    y_pred = model.predict(X_test_scaled)\n",
    "    corr_c.append(np.corrcoef(y_test, y_pred)[0, 1])\n",
    "\n",
    "    # 5) Tuning KernelRidgee + test correlation\n",
    "    lgbm = lgb.LGBMRegressor(\n",
    "        objective='regression',\n",
    "        learning_rate=0.05,\n",
    "        n_estimators=300,\n",
    "        num_leaves=31,\n",
    "        max_depth=6,\n",
    "        min_data_in_leaf=50,\n",
    "        feature_fraction=0.7,\n",
    "        bagging_fraction=0.7,\n",
    "        bagging_freq=1,\n",
    "        n_jobs=-1,\n",
    "        verbose=-1\n",
    "    )\n",
    "    lgbm.fit(X_train_scaled, y_train)\n",
    "    y_pred = lgbm.predict(X_test_scaled)\n",
    "    corr_c_t.append(np.corrcoef(y_test, y_pred)[0, 1])\n",
    "\n",
    "summary = pd.DataFrame({\n",
    "    \"gene_id\": good_gene_id,\n",
    "    \"max_peak_corr\": corr_max,\n",
    "    \"y_train_var\": y_train_var,\n",
    "    \"y_test_var\": y_test_var,\n",
    "    \"ridge_test_r\": corr_c,\n",
    "    \"lgbm_test_r\": corr_c_t,\n",
    "})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b2c855a4-d69c-4b4e-b81d-11a0d504e253",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>max_peak_corr</th>\n",
       "      <th>y_train_var</th>\n",
       "      <th>y_test_var</th>\n",
       "      <th>ridge_test_r</th>\n",
       "      <th>lgbm_test_r</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000115977</td>\n",
       "      <td>0.071275</td>\n",
       "      <td>6.815775</td>\n",
       "      <td>6.795776</td>\n",
       "      <td>0.139281</td>\n",
       "      <td>0.166833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ENSG00000107331</td>\n",
       "      <td>0.056327</td>\n",
       "      <td>4.622707</td>\n",
       "      <td>4.605511</td>\n",
       "      <td>0.114302</td>\n",
       "      <td>0.135594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>ENSG00000115657</td>\n",
       "      <td>0.076905</td>\n",
       "      <td>1.428170</td>\n",
       "      <td>1.467709</td>\n",
       "      <td>0.148674</td>\n",
       "      <td>0.146216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>ENSG00000108846</td>\n",
       "      <td>0.084350</td>\n",
       "      <td>0.644879</td>\n",
       "      <td>0.687800</td>\n",
       "      <td>0.137206</td>\n",
       "      <td>0.117925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>ENSG00000125257</td>\n",
       "      <td>0.112532</td>\n",
       "      <td>8.042895</td>\n",
       "      <td>8.034110</td>\n",
       "      <td>0.283323</td>\n",
       "      <td>0.291517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>ENSG00000136379</td>\n",
       "      <td>0.061913</td>\n",
       "      <td>2.587219</td>\n",
       "      <td>2.633019</td>\n",
       "      <td>0.102538</td>\n",
       "      <td>0.103657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>ENSG00000099204</td>\n",
       "      <td>0.073839</td>\n",
       "      <td>6.154472</td>\n",
       "      <td>6.195627</td>\n",
       "      <td>0.136348</td>\n",
       "      <td>0.136272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>ENSG00000175164</td>\n",
       "      <td>0.050723</td>\n",
       "      <td>5.164624</td>\n",
       "      <td>5.191978</td>\n",
       "      <td>0.102156</td>\n",
       "      <td>0.107978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>ENSG00000166016</td>\n",
       "      <td>0.056345</td>\n",
       "      <td>2.810173</td>\n",
       "      <td>2.714567</td>\n",
       "      <td>0.138496</td>\n",
       "      <td>0.166022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>ENSG00000225792</td>\n",
       "      <td>0.065270</td>\n",
       "      <td>2.034776</td>\n",
       "      <td>1.949889</td>\n",
       "      <td>0.115125</td>\n",
       "      <td>0.105529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>ENSG00000265206</td>\n",
       "      <td>0.070217</td>\n",
       "      <td>6.259258</td>\n",
       "      <td>6.303818</td>\n",
       "      <td>0.129724</td>\n",
       "      <td>0.132609</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_id  max_peak_corr  y_train_var  y_test_var  ridge_test_r  \\\n",
       "13   ENSG00000115977       0.071275     6.815775    6.795776      0.139281   \n",
       "34   ENSG00000107331       0.056327     4.622707    4.605511      0.114302   \n",
       "48   ENSG00000115657       0.076905     1.428170    1.467709      0.148674   \n",
       "56   ENSG00000108846       0.084350     0.644879    0.687800      0.137206   \n",
       "57   ENSG00000125257       0.112532     8.042895    8.034110      0.283323   \n",
       "87   ENSG00000136379       0.061913     2.587219    2.633019      0.102538   \n",
       "102  ENSG00000099204       0.073839     6.154472    6.195627      0.136348   \n",
       "105  ENSG00000175164       0.050723     5.164624    5.191978      0.102156   \n",
       "113  ENSG00000166016       0.056345     2.810173    2.714567      0.138496   \n",
       "224  ENSG00000225792       0.065270     2.034776    1.949889      0.115125   \n",
       "240  ENSG00000265206       0.070217     6.259258    6.303818      0.129724   \n",
       "\n",
       "     lgbm_test_r  \n",
       "13      0.166833  \n",
       "34      0.135594  \n",
       "48      0.146216  \n",
       "56      0.117925  \n",
       "57      0.291517  \n",
       "87      0.103657  \n",
       "102     0.136272  \n",
       "105     0.107978  \n",
       "113     0.166022  \n",
       "224     0.105529  \n",
       "240     0.132609  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4543803-4e82-462f-9adc-aba0ac92de4c",
   "metadata": {},
   "source": [
    "It "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "0b829780-2bf2-457c-a2be-9386bdd72339",
   "metadata": {},
   "outputs": [],
   "source": [
    "#I am going to do train/test split only once and for each new batch just reuse indexes.\n",
    "train_idx = X_train.index\n",
    "test_idx = X_test.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "d45db299-3836-494b-8210-212eae611e11",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "84753 21189 84753 21189\n"
     ]
    }
   ],
   "source": [
    "print(len(train_idx),len(test_idx),len(y_train),len(y_test))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "c2b8dce0-ff72-4132-9f4f-a9fecde4151f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "5c02e442-3ad2-4201-a680-07fbb7d9060b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X_train shape: (84753, 66)\n",
      "X_test shape: (21189, 66)\n",
      "y_train var: 0.06155252\n",
      "y_test var: 0.06528162\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "38a714c6-8ac8-4e96-86f8-79f5762bd3ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-0.007252602796665639\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d58aecda-e748-4f74-b2fd-88cbb67f958a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import make_scorer\n",
    "import numpy as np\n",
    "\n",
    "def pearsonr_score(y_true, y_pred):\n",
    "    return np.corrcoef(y_true, y_pred)[0,1]\n",
    "\n",
    "pearson_scorer = make_scorer(pearsonr_score, greater_is_better=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6a4624c-34be-403a-ad47-c6e994d5340e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc1d35f2-8cc9-48d8-b75e-14880b94cf45",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb44923d-4b08-4feb-886b-97cb657cac74",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4f595876-dc8f-4de3-b100-aacd257e42a2",
   "metadata": {},
   "source": [
    "Do not fogret to deal with gene_id's not present in external file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "73471531-73d7-4902-a80c-442cb6c61922",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['ENSG00000198695', 'ENSG00000198712', 'ENSG00000198727',\n",
      "       'ENSG00000198763', 'ENSG00000198786', 'ENSG00000198804',\n",
      "       'ENSG00000198840', 'ENSG00000198886', 'ENSG00000198888',\n",
      "       'ENSG00000198899', 'ENSG00000198938', 'ENSG00000212907',\n",
      "       'ENSG00000228253', 'ENSG00000278704'],\n",
      "      dtype='object', name='gene_id')\n",
      "14\n"
     ]
    }
   ],
   "source": [
    "print(missing)\n",
    "print(len(missing))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "994c9449-319f-4c1a-9845-ff3d544dcba1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
